### ISSUE TITLE
Marking integer inputs to a compiled region requires `torch._dynamo.assume_static_by_default = True`

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#141594]
From some discussion with @drisspg after looking https://github.com/pytorch/pytorch/issues/141486#issuecomment-2501574320.

Here is a slightly modified repro:

```
import torch
from torch.nn.attention.flex_attention import create_block_mask, flex_attention
flex_attention_compile = torch.compile(flex_attention, fullgraph=True)
fast_create = torch.compile(create_block_mask, fullgraph=True)
B = 1
S = 10_000
H = 8
D = 16
def sliding_window_mask(window_size: int):
    def mask_fn(b, h, q_idx, kv_idx):
        return (q_idx - kv_idx <= window_size // 2) & (kv_idx - q_idx <= window_size // 2)
    return mask_fn
mask_mod = sliding_window_mask(1024)
xs = [torch.randn(B, H, i, D, dtype=torch.float16, device="cuda", requires_grad=True) for i in range(10_000, 100_000, 1_000)]
# Warmpup
#block_mask = fast_create(mask_mod, B=None, H=None, Q_LEN=S, KV_LEN=S, device="cuda")
#torch._dynamo.mark_dynamic(xs[0], -2)
#flex_attention_compile(xs[0], xs[0], xs[0], block_mask=block_mask)
from pathlib import Path
torch.cuda.synchronize()
# with profiler(Path("user_2.json")):
from contextlib import nullcontext
with nullcontext():
    for x in xs:
        block_mask = fast_create(mask_mod, B=None, H=None, Q_LEN=x.size(-2), KV_LEN=x.size(-2), device="cuda")
        out = flex_attention_compile(x, x, x, block_mask=block_mask)
        bwd = out.sum().backward()
```

Running with `TORCH_LOGS="recompiles"` gives you an initial recompile with this message:
```
    triggered by the following guard failure(s):
    - 0/0: L['Q_LEN'] == 10000
```

What's happening is that:

(1) even though we mark `x.size(2)` has having a dynamic shape, we end up passing it **directly** as an input to the compiled region (`block_mask`)

(2) We don't have a way (that I could find) to mark specific integer inputs to a compiled region as dynamic. I found `torch._dynamo.assume_static_by_default=True`, but this flag looks deprecated, and also applies to all inputs.

One side note from talking to @drisspg is that it would be very nice if in eager mode, calling `Q_LEN=x.size(-2)` "remembered" that `Q_LEN` is now dynamic as well, without the user needing to do extra work to tell the compiler about the dynamism though. This would require doing the provenance tracking in eager though, which will have a performance cost.

**Comment by ezyang** [comment]
This is similar to https://github.com/pytorch/pytorch/issues/129623

But this is actually very irritating fix for real because in the code the user would like to write, we don't really want to force the user to say `Q_LEN=DynamicInt(x.size(-2))`. Who would know to do that? But neither is it convenient to make `x.size(-2)` return something special once x is marked dynamic.

In a fullgraph=True setting, we can potentially have an alternate API where we specify which "inputs" to the compiled region are dynamic. But this too is an irritating extra step the user has to do.

With PGO, if you have a profile, we can automatically mark it dynamic. But you have to generate the profile first.

Another idea is that we unspec the integer by default, see if it gets unified with something that is explicitly marked dynamic, and if it isn't, we spec it. This would be 100% automatic and have the desired semantics (without an extra pre-run to generate the profile, as in PGO).

**Comment by ezyang** [comment]
Because making a dim dynamic when it actually could have been static results in a very concrete performance loss (in fact, the current benchmarks we do measure precisely this, since we don't actually ever run the benchmark at different sizes.)

**Comment by ysiraichi** [comment]
That makes sense. I will start working on this idea of unspecializing int inputs. 
Let me know if we need more discussion.

**Comment by ezyang** [comment]
Note that unspec'ing int inputs and then respecializing them later is a good way to tank compile time, potentially. It's worth a try, at least as an optional config though. cc @laithsakka @bobrenjc93

**Comment by ysiraichi** [comment]
After looking into it, I don't think the solution you proposed solves this issue, specifically. It does get rid of `ConstraintViolationError` that unifies dynamic dims with int parameters, though. However, the issue here is a bit different: `fast_create` does **not** take tensor parameters. Thus, `Q_LEN` and `KV_LEN` will never get unified.

As far as I understand it, a solution to this issue is exactly #129623. I believe the unspec-spec step is not needed. In order to run the code snippet without recompilations, I think we need the following steps: 

- Wrap the dimension in `mark_dynamic` function
- Solve [#129623](https://github.com/pytorch/pytorch/issues/129623) by creating a `DynamicInt` class

Finally, the code snippet should be changed by wrapping the following elements with `DynamicInt` in the warm-up phase:

- `Q_LEN` and `KV_LEN` parameters of the `create_block_mask` function
- `block_mask` (output of the above function) shape elements

@ezyang Let me know what you think.

**Comment by ysiraichi** [comment]
On second thoughts, I believe we could replace the wrapping of `block_mask` attributes **(2nd bullet)** with the unspec-spec of integer variables. But, since `block_mask` is actually a custom class instance, we would need to unspecialize every int that goes through `VariableTracker.build`.

That said, all this might increase compilation time too much.

**Comment by ezyang** [comment]
Yes, my idea was I would unspec every int as well as every tensor size. And yes, compile time is a problem. It is worth trying though, and if we find hotspots @laithsakka can assess if we should fix those hotspots.
